site/app/controllers/grading/ImagesController.php

```diff
@@ -177,22 +182,36 @@ public function ajaxUploadImagesFiles() {
 
     /**
      * @Route("/courses/{_semester}/{_course}/flag_user_image", methods={"POST"})
-     * @AccessControl(role="INSTRUCTOR")
+     * @AccessControl(role="FULL_ACCESS_GRADER")
      */
-    public function flagUserImage() {
+    public function flagUserImage(): JsonResponse {
+        $user_id = $_POST['user_id'];
+
         if ($_POST['flag'] === 'true') {
             $new_state = 'flagged';
-            $success_msg = $_POST['user_id'] . '\'s image was successfully flagged.';
-            $fail_msg = 'Some error occurred flagging ' . $_POST['user_id'] . '\'s image.';
+            $icon_html = ImagesView::UNDO_ICON_HTML;
+            $href = "javascript:flagUserImage('$user_id', false)";
         }
         else {
             $new_state = 'preferred';
-            $success_msg = $_POST['user_id'] . '\'s image was successfully unflagged.';
-            $fail_msg = 'Some error occurred unflagging ' . $_POST['user_id'] . '\'s image.';
+            $icon_html = ImagesView::FLAG_ICON_HTML;
+            $href = "javascript:flagUserImage('$user_id', true)";
+        }
+
+        $result = $this->core->getQueries()->updateUserDisplayImageState($user_id, $new_state);
+        $user = $this->core->getQueries()->getSubmittyUser($user_id);
+        $display_image = $user->getDisplayImage();
+
+        if ($result && $display_image) {
+            return JsonResponse::getSuccessResponse([
+                'first_last_username' => $user->getDisplayedFirstName() . ' ' . $user->getDisplayedLastName(),
+                'image_data' => $display_image->getImageBase64MaxDimension(ImagesView::IMG_MAX_DIMENSION),
+                'image_mime_type' => $display_image->getMimeType(),
+                'icon_html' => $icon_html,
+                'href' => $href
```

Comment:
```suggestion
        if ($result) {
            return JsonResponse::getSuccessResponse([
                'first_last_username' => $user->getDisplayedFirstName() . ' ' . $user->getDisplayedLastName(),
                'image_data' => !is_null($display_image) ? $display_image->getImageBase64MaxDimension(ImagesView::IMG_MAX_DIMENSION) : "",
                'image_mime_type' => !is_null($display_image) ? $display_image->getMimeType() : "",
                'icon_html' => $icon_html,
                'href' => $href
```

Reply:
This shouldn't be needed.  If `$display_image` is `null` then it will be coerced to `false` when used in the conditional, there by canceling the success response and returning an error response.

Reply:
Actually, cancelling the success response and returning an error response whenever you try to flag a photo requires you to hit page refresh to reflect those changes...

Reply:
Changed